### PR TITLE
SOW-24: handle stages without dates and events not found

### PR DESCRIPTION
**Pull request by vferraro-scottlogic**
do not display stages that don't have dates

### CODE REVIEW

File: lib/timetable-visualisation/PlanViewer.tsx [hunk 1]
```diff
@@ -48,28 +48,42 @@ export const PlanViewer = ({
...
-        }
-        return {
-          name: stage.title,
-          startEvent,
-          endEvent,
-        };
-      }),
+          if (!startEvent) {
+            throw new Error("event not found");
+          }
```
Comment (Ro4052) [inline lib/timetable-visualisation/PlanViewer.tsx:80]: So I think another change we want is to not throw an error if we don't find an event?

File: lib/timetable-visualisation/PlanViewer.tsx [hunk 2]
```diff
@@ -140,9 +154,11 @@ export const PlanViewer = ({
               </th>
               <td className="govuk-table__cell govuk-!-padding-top-6">
                 {endEvent?.eventDate
-                  ? `${toDefaultLocalDateString(
```
[the remark this answers is not in the document]
Comment (Ro4052) [reply lib/timetable-visualisation/PlanViewer.tsx:143]: I'm finding this ternary a little confusing, I'll wait until you refactor and do a re-review 👍

[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply lib/timetable-visualisation/PlanViewer.tsx:143]: all done! ready to be re-reviewed :)

File: lib/timetable-visualisation/PlanViewer.tsx (hunk 1 above)
Comment (vferraro-scottlogic) [reply lib/timetable-visualisation/PlanViewer.tsx]: I have done a refactor of the whole thing. Now it should handle missing events as well as handling events with empty eventDate

File: lib/timetable-visualisation/PlanViewer.tsx [hunk 3]
```diff
@@ -46,33 +47,38 @@ export const PlanViewer = ({
...
         );
 
         const endEvent = timetableEvents.find(
-          (event) => event.developmentPlanEvent === stage.endEventKey
+          (event) => event.developmentPlanEvent === currentStage.endEventKey
         );
 
-        if (!startEvent) {
-          throw new Error("event not found");
+        if (startEvent && startEvent.eventDate) {
```
Comment (vferraro-scottlogic) [inline lib/timetable-visualisation/PlanViewer.tsx:61]: Note: for a stage to be shown we need to have an event matching the stage startEvent and the event has to have an eventDate. This means that we are going to display events with startDate and no endDate but not events with endDate and no startDate

Comment (Ro4052) [reply lib/timetable-visualisation/PlanViewer.tsx:61]: Cool, seems somewhat unlikely that someone would put an end date without a start (and we can always review later)

File: lib/timetable-visualisation/PlanViewer.tsx [hunk 4]
```diff
@@ -46,33 +47,38 @@ export const PlanViewer = ({
...
+      },
+      []
+    );
+
+    if (publishedEvent && publishedEvent.eventDate) {
+      foundStages.push({
+        name: "Local Development Scheme Published",
+        startEvent: publishedEvent,
+      });
+    }
```
Comment (Ro4052) [inline lib/timetable-visualisation/PlanViewer.tsx:78]: Has this event deliberately been moved to the bottom of the visualisation?

Comment (Ro4052) [reply lib/timetable-visualisation/PlanViewer.tsx:78]: Also, to note, we're still throwing on line 47 if the LDS published event isn't present

Comment (vferraro-scottlogic) [reply lib/timetable-visualisation/PlanViewer.tsx:78]: no actually I forgot that we want to display that one first. I will move it to the top 👍

Comment (vferraro-scottlogic) [reply lib/timetable-visualisation/PlanViewer.tsx:78]: and I will remove the throwing there as well

Comment (vferraro-scottlogic) [reply lib/timetable-visualisation/PlanViewer.tsx]: done! :)

**Review by Ro4052 — APPROVED** [review]
Looks good, can we update the PR title please?

**Pull request merged**